### PR TITLE
Add Prettier and Eslint linting and apply it.

### LINKED ISSUE #59: add automated linting & prettier formatting
see #57 and #49 and #55 ... #58 will take care of documentation

### PR DESCRIPTION
**Pull request by jhoover4**
Adding prettier and eslint for repo linting.

## Prettier

Prettier is a code formatter used by the javascript community to automatically enforce one code style so there is no arguing about "spaces/tabs" or single quotes or double quotes, etc. It leaves what we call "bike shedding" (arguing about dumb pointless things) at the door. See https://prettier.io/ for more details.

## Eslint

Eslint is used to check code for common errors and automatically correct them. This keeps it easy for reviewers to focus on more complex issues and helps you quickly fix issues. See https://eslint.org/ for more details. We're using the React extension here, which is a common extension for react apps.

## Usage

To use these tools run `npm run lint` for eslint or `npm run format` for prettier or their "check" versions which doesn't apply fixes, but only throws warnings or errors.

### CODE REVIEW

File: src/components/Nav/nav.js [hunk 1]
```diff
@@ -1,108 +1,118 @@
...
-import './nav.css';
-import Logo from '../Logo/Logo';
-import $ from 'jquery';
+import React from "react";
+import "./nav.css";
+import Logo from "../Logo/Logo";
+import $ from "jquery";
 
-// export default function Nav() {
+// eslint-disable-next-line react/prop-types
```
Comment (jhoover4) [inline src/components/Nav/nav.js]: This is how you disable checks. We'll prob need to deal with this at some point or disable it globally.

File: src/components/Nav/nav-BASE.js [hunk 2]
```diff
@@ -1,90 +1,113 @@
-import React from 'react';
+ /* eslint-disable */
```
Comment (jhoover4) [inline src/components/Nav/nav-BASE.js]: This file seemed like a draft so I disabled the whole check here

File: .eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  settings: {
+    react: {
+      version: "detect",
+    },
+  },
+  env: {
+    browser: true,
+    es2021: true,
+  },
+  extends: ["eslint:recommended", "plugin:react/recommended", "prettier"],
```
Comment (ProsperousHeart) [inline .eslintrc.js]: Do you by chance have a link about the prettier extension? I was able to find documentation on the other 2 as per #58 but haven't been able to find more on how this prettier extension works for ESLint

File: .eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+module.exports = {
+  settings: {
+    react: {
+      version: "detect",
+    },
+  },
+  env: {
+    browser: true,
+    es2021: true,
```
Comment (ProsperousHeart) [inline .eslintrc.js]: If we are using latest ECMA in line 14, why are we not also including it here? Saw [this](https://eslint.org/docs/latest/use/configure/language-options#specifying-environments). Do we need to constantly be updating this environment variable as ECMA progresses?

File: .prettierignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+build
+coverage
```
Comment (ProsperousHeart) [inline .prettierignore:2]:
```suggestion
coverage

# Ignore all HTML files:
*.html

```

Is there a reason this was not included? This is from the site you shared in #49 .

File: .eslintrc.js (hunk 3 above)
Comment (ProsperousHeart) [reply .eslintrc.js]: As per [here](https://github.com/codefordallas/codefordallas.github.io/pull/49/files#r1225452267) you said because we are using this file that the following were removed from the `package.json` file:
- `react-app`
- `react-app/jest`

I assume the react plugin takes care of the 1st, what about Jest? I remembered Jest being a separate install & configuration from React.

**NOTE:** this also has a comment in the related file

File: package.json [hunk 6]
```diff
@@ -19,13 +19,11 @@
     "start": "react-scripts start",
     "build": "react-scripts build",
     "test": "react-scripts test",
-    "eject": "react-scripts eject"
-  },
-  "eslintConfig": {
-    "extends": [
-      "react-app",
-      "react-app/jest"
```
Comment (ProsperousHeart) [inline package.json:27]: These were not added into the ESLint config file - [here](https://github.com/codefordallas/codefordallas.github.io/pull/49/files#r1225452267) you said because we are using this eslint config file that the section was removed, yet I don't see them represented in it:
- `react-app`
- `react-app/jest`

I assume the react plugin takes care of the 1st, what about Jest? I remembered Jest being a separate install & configuration from React.

File: package-lock.json [hunk 7]
```diff
@@ -15936,16 +15969,16 @@
       }
     },
     "node_modules/typescript": {
-      "version": "5.0.4",
-      "resolved": "https://registry.npmjs.org/typescript/-/typescript-5.0.4.tgz",
-      "integrity": "sha512-cW9T5W9xY37cc+jfEnaUvX91foxtHkza3Nw3wkoF4sSlKn0MONdkdEndig/qPBWXNkmplh3NzayQzCiHM4/hqw==",
+      "version": "4.9.5",
+      "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.9.5.tgz",
+      "integrity": "sha512-1FXk9E2Hm+QzZQ7z+McJiHL4NW1F2EzMu9Nq9i3zAaGqibafqYwCVU6WyWAuyQRRzOlxou8xZSyXLEN8oKj24g==",
```
Comment (ProsperousHeart) [inline package-lock.json:15973]: Why did we go down in this version?

File: package-lock.json [hunk 8]
```diff
@@ -15936,16 +15969,16 @@
...
       "peer": true,
       "bin": {
         "tsc": "bin/tsc",
         "tsserver": "bin/tsserver"
       },
       "engines": {
-        "node": ">=12.20"
+        "node": ">=4.2.0"
```
Comment (ProsperousHeart) [inline package-lock.json:15980]: why did we go so far down in this version?

File: src/App.js [hunk 9]
```diff
@@ -1,26 +1,26 @@
 // import logo from './logo.svg';
-import './App.css';
-import { useState } from 'react';
-import { useEffect } from 'react';
-import { Waypoint } from 'react-waypoint';
+import React  from 'react';
```
Comment (ProsperousHeart) [inline src/App.js]: Why is this added? It's not needed in the file.

File: src/App.js [hunk 10]
```diff
@@ -121,7 +143,9 @@ function App() {
         // console.log(`WP${wpNum} ${wpType} (${loc}) next (uncertain if updated): ${nextActiveNav}`);
         // console.log(nextActiveNav);
 
-        console.log(`SEtting NAv classes with ${intPrevNavStr} and ${intActiveNavStr}`)
+        console.log(
+          `SEtting NAv classes with ${intPrevNavStr} and ${intActiveNavStr}`
```
Comment (ProsperousHeart) [inline src/App.js:146]:
```suggestion
          `Setting Nav classes with ${intPrevNavStr} and ${intActiveNavStr}`
```
Something I noticed should have been changed prior.

File: src/components/About/About.js [hunk 11]
```diff
@@ -1,42 +1,43 @@
...
-                If you're interested in joining Code for Dallas, get in touch with us through
-                our contact page, or come to one of our monthly meetups. We welcome people
-                of all skill levels and backgrounds!
-            </p>
-        </section>
-    )
-}
+      <h3>How to Join</h3>
+      <p>
+        If you&apos;re interested in joining Code for Dallas, get in touch with us
```
Comment (ProsperousHeart) [inline src/components/About/About.js]: Why was this changed? I assume through the linter but this seems odd to me since it's the HTML file.

File: src/components/Mission/index.jsx [hunk 12]
```diff
@@ -21,8 +21,11 @@ class Mission extends Component {
   render() {
     return (
       <Wrapper>
-          <Heading>Mission Statement</Heading>
-          <Text>We're just trying to make the world a better place and come up with great mission statements!</Text>
+        <Heading>Mission Statement</Heading>
+        <Text>
+          We&apos;re just trying to make the world a better place and come up with
```
Comment (ProsperousHeart) [inline src/components/Mission/index.jsx]: Making note of another apostrophe replaced with ASCII. **About** had similar note.

File: src/components/Nav/nav-BASE.js (hunk 2 above)
Comment (ProsperousHeart) [reply src/components/Nav/nav-BASE.js]: I believe it may have been a file I was working with as a "base" when I was having issues. Likely will be removed with #45 or a future issue for code cleanup

File: src/components/Nav/nav.js (hunk 1 above)
Comment (ProsperousHeart) [reply src/components/Nav/nav.js]: Added to documentation for #58

File: src/components/Projects/chatGPT-Projects.html [hunk 13]
```diff
@@ -83,4 +83,3 @@ <h3>Project 3</h3>
   crossorigin="anonymous"></script>
   
 </body>
```
Comment (ProsperousHeart) [inline src/components/Projects/chatGPT-Projects.html:104]:
```suggestion
</body>
</html>
```
This was removed but the original file had the start of the HTML code. On the next go-round this file will be removed anyway.

**Review by ProsperousHeart — CHANGES_REQUESTED** [review]
A couple changes & a few clarifications requested

**Comment by ProsperousHeart** [comment]
Issue #58 is the documentation outlining these changes for newbies. Taken care of under PR #60

File: .eslintrc.js (hunk 3 above)
Comment (jhoover4) [reply .eslintrc.js]: > Do you by chance have a link about the prettier extension? I was able to find documentation on the other 2 as per #58 but haven't been able to find more on how this prettier extension works for ESLint

Sure do! https://prettier.io/docs/en/integrating-with-linters.html

Comment (jhoover4) [reply .eslintrc.js]: > As per [here](https://github.com/codefordallas/codefordallas.github.io/pull/49/files#r1225452267) you said because we are using this file that the following were removed from the `package.json` file:
> 
> - `react-app`
> 
> - `react-app/jest`
> 
> 
> 
> I assume the react plugin takes care of the 1st, what about Jest? I remembered Jest being a separate install & configuration from React.
> 
> 
> 
> **NOTE:** this also has a comment in the related file

Ah good point. I was planning on setting up jest and adding it into the CI in another PR. Didnt realize we already had it setup. I can revert if we want or just add it all in the test PR

File: .prettierignore (hunk 5 above)
Comment (jhoover4) [reply .prettierignore:2]: So prettier can run on html files if we want or we can have it exclude them. I think we might as well have it run for everything, but up to yall

File: package.json (hunk 6 above)
Comment (jhoover4) [reply package.json:27]: Yep I think I just missed that. See above

File: src/App.js (hunk 9 above)
Comment (jhoover4) [reply src/App.js]: Eslint was trigger on it, but I did some research and apparently its a legacy thing. See https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#eslint. Ill turn it off

File: src/App.js (hunk 10 above)
Comment (jhoover4) [reply src/App.js:146]: No worries!

File: src/components/About/About.js (hunk 11 above)
Comment (jhoover4) [reply src/components/About/About.js]: Another eslint trigger. Here its because you shouldnt use literal characters in html. Its safer to use the character code. In this case its a "'". I think we should keep this one on :)

File: src/components/Mission/index.jsx (hunk 12 above)
Comment (jhoover4) [reply src/components/Mission/index.jsx]: Yep!

File: src/components/Projects/chatGPT-Projects.html (hunk 13 above)
Comment (jhoover4) [reply src/components/Projects/chatGPT-Projects.html:104]: Yeah eslint was complaining here because technically you dont need this closing tag as the first one is self closing.

File: package-lock.json (hunk 7 above)
Comment (jhoover4) [reply package-lock.json:15973]: Whoops!! Looks like a mistake on my part good catch Ill fix it

**Comment by jhoover4** [comment]
FYI any changes I made to /src came as a response to eslint error or prettier formatting. Looks like there a few rules we can disable, but just wanted to give you a heads up that that is where the changes came from.

File: .eslintrc.js (hunk 4 above)
Comment (jhoover4) [reply .eslintrc.js]: I don't believe so, but this was actually generated by eslint itself so I'm hesitant about changing it. I'm reading up on it in https://eslint.org/docs/latest/use/configure/language-options

**Comment by jhoover4** [comment]
OK. So I realized I messed up the eslint config a bit cause I wasn't letting it play nice with create react app. I fixed that per the docs here: https://create-react-app.dev/docs/setting-up-your-editor/#extending-or-replacing-the-default-eslint-config. Might want to include that in your notes as it also has setups for IDE/code editors (mine is already working with WebStorm).

File: .eslintrc.js [hunk 14]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  extends: ["react-app", "plugin:jsx-a11y/recommended", "prettier"],
+  plugins: ["jsx-a11y"],
```
Comment (jhoover4) [inline .eslintrc.js:3]: Added this as week as I think having a11y checks is something we'd want knowing our audience.

File: package.json (hunk 6 above)
Comment (jhoover4) [reply package.json:27]: Fixed it

File: .eslintrc.js (hunk 3 above)
Comment (ProsperousHeart) [reply .eslintrc.js]: > I was planning on setting up jest and adding it into the CI in another PR. Didnt realize we already had it setup. I can revert if we want or just add it all in the test PR

If you're just going to add it in the test PR, then I'm fine pushing this as is.

> > Do you by chance have a link about the prettier extension? I was able to find documentation on the other 2 as per #58 but haven't been able to find more on how this prettier extension works for ESLint
> 
> Sure do! https://prettier.io/docs/en/integrating-with-linters.html

As to the link, thank you. That's the one I saw - I'll try to remember to link ones I find in the future and be more clear about the ask. I was looking more for [this](https://github.com/prettier/eslint-plugin-prettier#installation) (which I overlooked before I inquired - my bad)

File: .prettierignore (hunk 5 above)
Comment (ProsperousHeart) [reply .prettierignore:2]: Got it. That should be fine - we can always add it back in if needed.

I was only curious because it was in the documentation

File: src/App.js (hunk 9 above)
Comment (ProsperousHeart) [reply src/App.js]: Awesome! Also added this to documentation in resources section for #58 related to #60

File: src/components/Projects/chatGPT-Projects.html (hunk 13 above)
Comment (ProsperousHeart) [reply src/components/Projects/chatGPT-Projects.html:104]: *smacks forehead* yup - thanks for the reminder

**Comment by ProsperousHeart** [comment]
> OK. So I realized I messed up the eslint config a bit cause I wasn't letting it play nice with create react app. I fixed that per the docs here: https://create-react-app.dev/docs/setting-up-your-editor/#extending-or-replacing-the-default-eslint-config. Might want to include that in your notes as it also has setups for IDE/code editors (mine is already working with WebStorm).

Added [here](https://github.com/codefordallas/codefordallas.github.io/pull/60/commits/5456dfe643dd9129a16ead369946955429367d63)

File: src/components/Projects/chatGPT-Projects.html [hunk 15]
```diff
@@ -1,47 +1,59 @@
...
+    name="description"
+    property="og:description"
+    content="Code For Dallas public site"
+  />
 
-    <title>Code For Dallas</title>
-    <meta property="og:title" content="Code For Dallas Hoem Page">
-    <meta name="author" content="Code For Dallas Volunteers">
+  <title>Code For Dallas</title>
+  <meta property="og:title" content="Code For Dallas Hoem Page" />
```
Comment (ProsperousHeart) [inline src/components/Projects/chatGPT-Projects.html]:
```suggestion
  <meta property="og:title" content="Code For Dallas Home Page" />
```
Misspelling that wasn't caught prior

File: .eslintrc.js (hunk 14 above)
Comment (ProsperousHeart) [inline .eslintrc.js]:
```suggestion
  extends: ["react-app", "prettier"],
  plugins: ["jsx-a11y"],
```
According to [this](https://www.npmjs.com/package/eslint-plugin-jsx-a11y) we don't need to do both `extends` and `plugin`. Is there a reason both are there that I was unable to locate?

image

I removed from extension so users could see both variables.

File: .eslintrc.js [hunk 16]
```diff
@@ -0,0 +1,4 @@
...
+};
```
Comment (ProsperousHeart) [inline .eslintrc.js:4]: I saw the `env` section was removed (from [here](https://github.com/codefordallas/codefordallas.github.io/pull/57#discussion_r1225943765)) - were these no longer needed?

File: package-lock.json (hunk 7 above)
Comment (ProsperousHeart) [reply package-lock.json:15973]: Adding additional note since you made changes but this was not one of them

UPDATE:  since it didn't include the original conversation on the **conversation** page ... TypeScript version went down from `5.0.4`

File: package-lock.json (hunk 8 above)
Comment (ProsperousHeart) [reply package-lock.json:15980]: Adding additional note since you made changes but this was not one of them (downgrade from node engine `>=12.20`)

File: .eslintrc.js (hunk 14 above)
Comment (ProsperousHeart) [reply .eslintrc.js:3]: updated in most recent push to #60

**Review by ProsperousHeart — CHANGES_REQUESTED** [review]
Updated documentation with latest changes in #60 PR

A few inquiries and changes requested

File: .eslintrc.js (hunk 16 above)
Comment (jhoover4) [reply .eslintrc.js:4]: No that would be defining it twice. If we want to define that now we could do it here.

File: package-lock.json (hunk 7 above)
Comment (jhoover4) [reply package-lock.json:15973]: Oh I didnt realize this was the lock file. You shouldn't really be checking this, its autogenerated and just so environments are the same. For example, this is a transitive dep and we don't really want to control it directly.

File: package-lock.json (hunk 8 above)
Comment (jhoover4) [reply package-lock.json:15980]: I wouldn't worry about this. See comment above. If you want to pin the node version we can do that in package.json

File: .eslintrc.js (hunk 16 above)
Comment (ProsperousHeart) [reply .eslintrc.js:4]: if you think it's better to do it the other way (as is now in the code) then I'm good with keeping it like this

File: .eslintrc.js (hunk 4 above)
Comment (ProsperousHeart) [reply .eslintrc.js]: For documentation purposes - this was addressed [here](https://github.com/codefordallas/codefordallas.github.io/pull/57#discussion_r1226816018)

**Review by ProsperousHeart — APPROVED** [review]
All items have been addressed, code reviewed, and successfully run locally. Approved! Thank you for your time and patience. Documentation for this has been updated under #60 - to be reviewed and integrated into branch `v1-mvp`
